rfcs/55-deterministic_rng.md

```diff
@@ -0,0 +1,155 @@
+To make sure the behavior works, you write a test to first check the computer's hand and if it is good confirm that all its money is bet.
+If the test passes does it ensure the computer behaves as intended? Sadly, no.
+
+Because the deck is randomly shuffled for each game--without doing so the player would already know the card order from the previous game--it is not guaranteed that the computer player gets a good hand and thus the betting logic goes unchecked.
+While there are ways around this--a fake deck that is not shuffled, running the test many times to increase confidence, breaking the logic into units and testing those--it would be very helpful to have randomness as well as a way to make it _less_ random.
+
+Luckily, when a computer needs a random number it doesn't use real randomness and instead uses a [pseudorandom number generator](https://en.wikipedia.org/wiki/Pseudorandom_number_generator).
+Popular Rust libraries containing pseudorandom number generators are [`rand`](https://crates.io/crates/rand) and [`fastrand`](https://crates.io/crates/fastrand).
+
+Pseudorandom number generators require a source of [entropy](https://en.wikipedia.org/wiki/Entropy) called a [random seed](https://en.wikipedia.org/wiki/Random_seed).
+The random seed is used as input to generate numbers that _appear_ random but are instead in a specific and deterministic order.
+For the same random seed, a pseudorandom number generator always returns the same numbers in the same order.
+
+For example, let's say you seed a pseudorandom number generator with `1234`.
+You then ask for a random number between `10` and `99` and the pseudorandom number generator returns `12`.
+If you run the program again with the same seed (`1234`) and ask for another random number between `1` and `99`, you will again get `12`.
+If you then change the seed to `4567` and run the program, more than likely the result will not be `12` and will instead be a different number.
+If you run the program again with the `4567` seed, you should see the same number from the previous `4567`-seeded run.
+
+There are many types of pseudorandom number generators each with their own strengths and weaknesses.
+Because of this, Bevy does not include a pseudorandom number generator.
+Instead, the `bevy_entropy` plugin includes a source of entropy to use as a random seed for your chosen pseudorandom number generator.
+
+Note that Bevy currently has [other sources of non-determinism](https://github.com/bevyengine/bevy/discussions/2480) unrelated to pseudorandom number generators.
+
+### Usage
+
+The `bevy_entropy` plugin ships with Bevy and is enabled by default.
+If you do not need randomness, you may [disable the plugin](https://docs.rs/bevy/latest/bevy/app/struct.PluginGroupBuilder.html#method.disable) or use Bevy's [minimal set of plugins](https://docs.rs/bevy/latest/bevy/struct.MinimalPlugins.html).
+
+When enabled, `bevy_entropy` provides one world resource: `Entropy`.
```

Comment:
Totally agree Bevy needs seed-able randomness built-in. :)

But I don't think providing a premade resource like this is good idea.
Most standard sized games have more than one RNG.
Take Minecraft for example, you have world randomness, you have mob randomness, etc.
Therefore providing an out of the box singleton could lead to bad practices.

Here's how I'd suggest going about it:
- The default `bevy_entropy` should provide traits for RNG types (wrappers or re-exports for randomness crates),
- The plugin should not be included in the `DefaultPlugins` and only added explicitly by the user,
- The plugin would add the `Entropy` as you call it,
- Both, the plugin and `Entropy` would be renamed to something that implies it's **not the only** source of randomness, something like `DebugEntropy` is good, because it implies you may want to remove it in a "full game" for  something you have more control over.

Then you end up with:
- Randomness provided through Bevy,
- Easily installable default seeded RNG for prototyping,
- Incentive to crate your own sources of RNG when finalizing a feature.

Reply:
With `bevy_rand`, the plugin is entirely based on instantiating the resource *and* component *types*, registering them for reflection purposes as well. It also is generic as it makes no assumption on the PRNG algorithm the user wants to make use of (given there are different PRNG kinds with different properties and trade-offs). The PRNG types are in `bevy_prng` in order to account for the reflection traits, etc.

What we get is a Global source of entropy (seedable as well), and component types that can be attached to entities to serve as multiple sources. We get forking between Resource -> Component and also Component -> Component, so with some consideration, we can seed many sources from a single seed in a deterministic but secure manner (no duplicate seeds/states). Sources can be seeded from a thread local, OS/hardware seed source for non-determinism, or from a specified seed.

Needs for a source can vary from the very simple (super simple games in Bevy can be done entirely with Resources anyway), to complex (such as needing strict deterministic seeding and generation across many sources). I would suggest that *how* people want to use RNG in games is not something we should be biasing one way or another, because how much people want to go down is entirely dependent on their needs. Should we care if they opt for a single global source if that's all they need? Maybe RNG is not used often, mostly during instantiation and it being non-deterministic is fine?

Ultimately, I'm not convinced about forcing "best practices" for games. For crates/plugins, I'm more convinced that there should be particular patterns for RNG sources, so that one plugin does not interfere with another. For games, using global sources is fine, because it will just go into a single binary. Crates however need to have their own sources and also expose the means to set the seeds for the sources they use.

Reply:
Good point about plugins, definitely more direct than the "good practices" argument, although I still think it's important.

Totally agree with the game dev's decision to use a single RNG, I just don't think it should be implied by the code.
When working with different libraries I often based my thought on the API provided, if the API suggested a solution I'd usually keep using it until it was absolutely necessary to find out better ways. That's why signaling something like this in the name itself could be helpful. (second thought, there *is* the description)
A counterargument can be made that only advanced devs will need to use multiple RNGs and they're knowledgeable enough to use them. But advanced devs got there somehow, and leaving tips that cost nothing around the code seems like the least we can do.

About the Resource -> Component, Component -> Component. I think what I'm missing is Resource -> Resource or just Seed -> Resource. I think that's probably covered too, but let me know if that's incorrect.

Reply:
Resource -> Resource forking is... I think not viable, or a very rare case, as Resources should be treated as a global source, and global sources should not depend on each other. I'd also encourage fewer global sources than lots of global sources, so then the need for Resource -> Resource forking is lessened. Better for Seed -> Resource, as that can be easily done. Global sources should be more generic and broad, not "granular" in ways that might suggest it has a purpose like "mob RNG" or "world RNG". If you want more granular/scoped RNG sources, that should be as a Component/Entity.

Most of this is *patterns* though, and the best way to really get devs to use these useful patterns is to provide documentation/examples. It's *hard* to enforce patterns via code/APIs.

Reply:
Yeah, Seed -> Resource should be enough, since the seed can come from another RNG, another resource, or some configuration/save file.

Good documentation and examples should be enough.